src/main/scala/com/advancedtelematic/campaigner/db/Campaigns.scala

```diff
@@ -165,22 +175,52 @@ protected [db] class Campaigns(implicit db: Database, ec: ExecutionContext)
         counts.getOrElse(DeviceStatus.failed, 0).toLong
     )
 
+    def processFailures(failedDevices: Set[DeviceUpdate], retryCampaigns: Set[Campaign]): Set[CampaignFailureStats] = {
+      val missingErrorCode = "MISSING_ERROR_CODE"
+      val retryStatusByFailureCode = retryCampaigns
+        .groupBy(_.failureCode.getOrElse(missingErrorCode))
+        .mapValues { campaigns =>
+          if (campaigns.exists(_.status == CampaignStatus.finished)) {
```

Comment:
Only one campaign here?

Reply:
For the current user story I think one campaign, yes. Because we can only launch one retry-campaign for failure code, hence grouping by failure code should be the same as grouping by campaign. If/when we generalise this, then we could have more than one campaign in each group.